http3/request.go

```diff
@@ -13,118 +13,147 @@ import (
 	"github.com/quic-go/qpack"
 )
 
-func requestFromHeaders(headers []qpack.HeaderField) (*http.Request, error) {
-	var path, authority, method, protocol, scheme, contentLengthStr string
+type header struct {
+	// Pseudo header fields defined in RFC 9114
+	Path      string
+	Method    string
+	Authority string
+	Scheme    string
+	Status    string
+	// for Extended connect
+	Protocol string
+	// parsed and deduplicated
+	ContentLength int64
+	// all non-pseudo headers
+	Headers http.Header
+}
 
-	httpHeaders := http.Header{}
+func parseHeaders(headers []qpack.HeaderField, isRequest bool) (header, error) {
+	hdr := header{Headers: make(http.Header, len(headers))}
 	var readFirstRegularHeader bool
+	var contentLengthStr string
 	for _, h := range headers {
 		// field names need to be lowercase, see section 4.2 of RFC 9114
 		if strings.ToLower(h.Name) != h.Name {
-			return nil, fmt.Errorf("header field is not lower-case: %s", h.Name)
+			return header{}, fmt.Errorf("header field is not lower-case: %s", h.Name)
 		}
 		if !httpguts.ValidHeaderFieldValue(h.Value) {
-			return nil, fmt.Errorf("invalid header field value for %s: %q", h.Name, h.Value)
+			return header{}, fmt.Errorf("invalid header field value for %s: %q", h.Name, h.Value)
 		}
 		if h.IsPseudo() {
 			if readFirstRegularHeader {
 				// all pseudo headers must appear before regular header fields, see section 4.3 of RFC 9114
-				return nil, fmt.Errorf("received pseudo header %s after a regular header field", h.Name)
+				return header{}, fmt.Errorf("received pseudo header %s after a regular header field", h.Name)
+			}
+			var isResponsePseudoHeader bool // pseudo headers are either valid for requests or for responses
+			switch h.Name {
+			case ":path":
+				hdr.Path = h.Value
+			case ":method":
+				hdr.Method = h.Value
+			case ":authority":
+				hdr.Authority = h.Value
+			case ":protocol":
+				hdr.Protocol = h.Value
+			case ":scheme":
+				hdr.Scheme = h.Value
+			case ":status":
+				hdr.Status = h.Value
+				isResponsePseudoHeader = true
+			}
+			if isRequest && isResponsePseudoHeader {
+				return header{}, fmt.Errorf("invalid request pseudo header: %s", h.Name)
+			}
+			if !isRequest && !isResponsePseudoHeader {
+				return header{}, fmt.Errorf("invalid response pseudo header: %s", h.Name)
```

Comment:
This doesn't fail if there is no `:status` in response. 

```
diff --git a/http3/request_test.go b/http3/request_test.go
index f6a1a424..319b66fb 100644
--- a/http3/request_test.go
+++ b/http3/request_test.go
@@ -100,6 +100,14 @@ var _ = Describe("Request", func() {
 		Expect(err).To(MatchError("invalid request pseudo header: :status"))
 	})
 
+	It("rejects response with no status field", func() {
+		headers := []qpack.HeaderField{
+			{Name: "content-length", Value: "42"},
+		}
+		_, err := parseHeaders(headers, false)
+		Expect(err).ToNot(BeNil())
+	})
+
 	It("parses path with leading double slashes", func() {
 		headers := []qpack.HeaderField{
 			{Name: ":path", Value: "//foo"},
```

Reply:
That's something that should go into `responseFromHeaders` in #3969, isn't it? Validation logic for required request headers is in `requestFromHeaders`, not in `parseHeader`.

Reply:
Okay. Makes sense. Absent required fields are checked in requestForHeaders. So this is inline with it. Sorry for blocking here.